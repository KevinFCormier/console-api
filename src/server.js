/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import https from 'https';
import fs from 'fs';
import schema from './schema/index';
import config from '../config';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const GRAPHQL_PORT = process.env.PORT || config.get('httpPort') || 4000;
const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(req => ({ formatError, schema, context: req })));

const privateKey = fs.readFileSync(process.env.serverKey || './sslcert/hcmuiapi.key', 'utf8');
const certificate = fs.readFileSync(process.env.serverCert || './sslcert/hcmuiapi.crt', 'utf8');
const credentials = { key: privateKey, cert: certificate };
const server = https.createServer(credentials, graphQLServer);

server.listen(GRAPHQL_PORT, () => {
  logger.info(`[pid ${process.pid}] [env ${process.env.NODE_ENV}] started.`);
  logger.info(`HCM UI API is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphql`);
  if (process.env.NODE_ENV !== 'production') {
    logger.info(`GraphiQL is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphiql`);
  }
});
