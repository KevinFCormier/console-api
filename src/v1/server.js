/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import schema from './schema/index';
import config from '../../config';
import HCMConnector from './datasource/lib/hcm-connector';
import MongoConnector from './datasource/lib/mongo-connector';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};


const getProductionOptions = async (req) => {
  const mongoConnector = new MongoConnector();
  const hcmConnector = new HCMConnector({
    hcmUrl: config.get('hcmUrl') || 'http://localhost:8080',
    pollInterval: config.get('hcmPollInterval') || 200,
    pollTimeout: config.get('hcmPollTimeout') || 10000,
  });

  await mongoConnector.connect(config.get('mongodbUrl') || 'mongodb://localhost:27017/weave');
  return {
    formatError,
    schema,
    context: {
      req,
      hcmConnector,
      mongoConnector,
    },
  };
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(getProductionOptions));

export default graphQLServer;
