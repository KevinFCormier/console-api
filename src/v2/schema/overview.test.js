/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Overview Resolver', () => {
  test('Correctly Resolves Overview Query', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
        overview {
          clusters {
            metadata {
              name
              namespace
              labels
              uid
            }
            capacity {
              cpu
              memory
              nodes
              storage
            }
            usage {
              cpu
              memory
              pods
              storage
            }
            status
          }
          applications {
            metadata {
              name
              namespace
            }
            raw
            selector
          }
          pods {
            metadata {
              name
              namespace
            }
            cluster {
              metadata {
                name
              }
            }
            status
          }
        }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
