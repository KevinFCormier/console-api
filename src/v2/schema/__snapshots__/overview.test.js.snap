// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview Resolver Correctly Resolves Overview Query 1`] = `
Object {
  "data": Object {
    "overview": Object {
      "applications": Array [
        Object {
          "metadata": Object {
            "name": "gbapp-gbapp",
            "namespace": "default",
          },
          "raw": Object {
            "metadata": Object {
              "annotations": Object {
                "applicationRelationship": "gbapp-gbapp-appfrontend,gbapp-gbapp-master,gbapp-gbapp-slave",
                "dashboard": "https://9.42.81.137:8443/grafana/dashboard/db/gbapp-gbapp-dashboard-via-federated-prometheus",
                "deployable": "gbapp-gbapp,gbapp-gbapp-redismaster,gbapp-gbapp-redisslave",
                "mcm.ibm.com/user-group": "c3lzdGVtOnNlcnZpY2VhY2NvdW50cyxzeXN0ZW06c2VydmljZWFjY291bnRzOmt1YmUtc3lzdGVtLHN5c3RlbTphdXRoZW50aWNhdGVk",
                "mcm.ibm.com/user-identity": "c3lzdGVtOnNlcnZpY2VhY2NvdW50Omt1YmUtc3lzdGVtOmRlZmF1bHQ=",
                "placementBinding": "gbapp-gbapp,gbapp-gbapp-redismaster",
              },
              "creationTimestamp": "2018-11-12T18:48:14Z",
              "labels": Object {
                "app": "gbapp",
                "chart": "gbapp-0.1.0",
                "heritage": "Tiller",
                "name": "gbapp-gbapp",
                "release": "gbapp",
              },
              "name": "gbapp-gbapp",
              "namespace": "default",
              "resourceVersion": "281",
              "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/default/applications/gbapp-gbapp",
              "uid": "8301b18b-e6ab-11e8-b778-9e17a0ceb481",
            },
            "spec": Object {
              "componentKinds": Array [
                Object {
                  "group": "core",
                  "kind": "Pods",
                },
              ],
              "descriptor": Object {},
              "selector": Object {
                "matchExpressions": Array [
                  Object {
                    "key": "app",
                    "operator": "In",
                    "values": Array [
                      "gbapp",
                      "gbf",
                      "gbrm",
                      "gbrs",
                    ],
                  },
                ],
              },
            },
            "status": Object {},
          },
          "selector": Object {
            "matchExpressions": Array [
              Object {
                "key": "app",
                "operator": "In",
                "values": Array [
                  "gbapp",
                  "gbf",
                  "gbrm",
                  "gbrs",
                ],
              },
            ],
          },
        },
      ],
      "clusters": Array [
        Object {
          "capacity": Object {
            "cpu": "18",
            "memory": "32013Mi",
            "nodes": 2,
            "storage": "100Gi",
          },
          "metadata": Object {
            "labels": Object {
              "cloud": "IBM",
              "datacenter": "toronto",
              "environment": "Dev",
              "name": "managed-cluster",
              "owner": "marketing",
              "region": "US",
              "vendor": "ICP",
            },
            "name": "managed-cluster",
            "namespace": "default",
            "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
          },
          "status": "ok",
          "usage": Object {
            "cpu": "6598m",
            "memory": "6728Mi",
            "pods": 5071,
            "storage": "60Gi",
          },
        },
        Object {
          "capacity": Object {
            "cpu": "16",
            "memory": "32013Mi",
            "nodes": 3,
            "storage": "80Gi",
          },
          "metadata": Object {
            "labels": Object {
              "cloud": "IBM",
              "datacenter": "raleigh",
              "environment": "Dev",
              "name": "hub-cluster",
              "owner": "development",
              "region": "US",
              "vendor": "ICP",
            },
            "name": "hub-cluster",
            "namespace": "kube-system",
            "uid": "d9f3a5aa-9f19-11e8-855e-f2b998610544",
          },
          "status": "ok",
          "usage": Object {
            "cpu": "6598m",
            "memory": "6728Mi",
            "pods": 5071,
            "storage": "60Gi",
          },
        },
      ],
      "pods": Array [
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "ibm-cert-manager-cert-manager-749ffc8bbc-grfrg",
            "namespace": "cert-manager",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "nginx",
            "namespace": "default",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "audit-logging-fluentd-ds-lgjpj",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "audit-logging-fluentd-ds-qzvth",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "audit-logging-fluentd-ds-v7r5b",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "auth-apikeys-rsfrx",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "auth-idp-nl5s5",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "auth-pap-dqsnz",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "auth-pdp-brpms",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "calico-kube-controllers-78d477cf58-82dhs",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "calico-node-fzxmv",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "calico-node-k8xb6",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "calico-node-mhz49",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "catalog-ui-jc5v5",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "custom-metrics-adapter-5dfb859b4b-2xr86",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
        Object {
          "cluster": Object {
            "metadata": Object {
              "name": "hub-cluster",
            },
          },
          "metadata": Object {
            "name": "default-backend-5c6c964bf5-fws56",
            "namespace": "kube-system",
          },
          "status": "Running",
        },
      ],
    },
  },
}
`;
